### PR TITLE
[1863] Operators True/False are ignored

### LINKED ISSUE #1863: Operators True/False are ignored
If object implements True/False operators, than operators || and && must be invoked this way:

left || right => True(Left) ? Left : Left | Right;
left && right => False(Left) ? Right : Left & Right;

In JS: 
left || right  => (var $t = left, op_True($t) ? $t : op_BitwiseOr($t, right))
left && right => (var $t = right, op_False($t) ? right : op_BitwiseAnd($t, right))

### Steps To Reproduce
https://dotnetfiddle.net/1CkOFx
http://deck.net/09e814f22472c9fc18eebdd938388107
```csharp
using System;
public class Obj
{
	private string _v;
	public string v{
		get
		{
			return _v;
		}
		set
		{
			_v = value;
		}
	}
    
	public Obj(string v)
	{
        this.v = v;
		Console.WriteLine("Create");
	}
    
    
    static public bool operator false(Obj o)
    {
		Console.WriteLine("False called");
        return o ==null;
    }
    
    
    static public bool operator true(Obj o)
    {
		Console.WriteLine("True called");
        return o !=null;
    }
    
    static public Obj operator |(Obj left, Obj right)
    {
        if(left == null)
            return right;
        
        if (right == null)
            return left;
        
        return left;
    }
    
    static public Obj operator &(Obj left, Obj right)
    {
         if(left == null)
            return null;
        
        if (right == null)
            return null;
        
        return right;
    }
    
    public override string ToString()
    {
        return v;
    }
}

public class Program
{
    public static void Main()
    {
        Console.WriteLine(">>>");
        //Expected: Bridge.Console.log((var $t = new Obj("left"), Obj.op_True($t) ? $t : Obj.op_BitwiseOr($t, new Obj("right")))
		Console.WriteLine(new Obj("left") || new Obj("right"));
        //Expected: Bridge.Console.log((var $t = new Obj("left"), Obj.op_False($t) ? new Obj("right") : Obj.op_BitwiseAnd($t, new Obj("right")))
		Console.WriteLine(new Obj("left") && new Obj("right"));
        Console.WriteLine("<<<");
    }
}
```

### PR DESCRIPTION
**Pull request by vladsch**
Fixes #1863

### CODE REVIEW

**Comment by AndreyZM** [comment]
its probably produce invalid code:
```csharp
       private void answerGroup(Object data)
        {
            dispatchEvent(new  QuestEvent(QuestEvent.GROUP, data && data.length > 0 && (data[0]["subscribe"] == 1)));
        }
```

```js
answerGroup: function (data) {
            var $t;
            this.dispatchEvent(new com.api.QuestEvent.$ctor1(com.api.QuestEvent.GROUP, as3.Object.op_Implicit$10(($t = as3.Object.op_False(data) ? data : as3.Object.op_BitwiseAnd(data, as3.Object.op_Implicit$2(data.getlength() > 0))), as3.Object.op_False($t) ? $t : as3.Object.op_BitwiseAnd($t, as3.Object.op_Implicit$2((as3.Object.op_Implicit$13(data.getItem(0).getItem("subscribe")) === 1)))))));
        },
```

**Comment by vladsch** [comment]
Can you provide test case which reproduces the issue?

**Comment by AndreyZM** [comment]
```csharp
using System;
public class Obj
{
    private object value;
    public int length { get; set; }

    public Obj()
    {
    }


    static public bool operator false(Obj o)
    {
        return o == null;
    }


    static public bool operator true(Obj o)
    {
        return o != null;
    }

    static public implicit operator Obj(bool v)
    {
        return new Obj() { value = v };
    }

    static public implicit operator bool(Obj v)
    {
        return v.value.As<bool>();
    }

    static public implicit operator int(Obj v)
    {
        return v.value.As<int>();
    }

    public Obj this[object key] { get { return null; } set { } }

    static public implicit operator Obj(int v)
    {
        return new Obj() { value = v };
    }

    static public Obj operator |(Obj left, Obj right)
    {
        if (left == null)
            return right;

        if (right == null)
            return left;

        return left;
    }

    static public Obj operator &(Obj left, Obj right)
    {
        if (left == null)
            return null;

        if (right == null)
            return null;

        return right;
    }

    public override string ToString()
    {
        return value.ToString();
    }
}

public class Program
{
    public static void Main()
    {
        var data = new Obj();
        Method(data && data.length > 0 && (data[0]["key"] == 1));
    }

    public static void Method(bool b)
    {

    }
}
```

File: Compiler/Translator/Emitter/Blocks/BinaryOperatorBlock.cs [hunk 1]
```diff
@@ -52,6 +65,66 @@ protected bool ResolveOperator(BinaryOperatorExpression binaryOperatorExpression
...
+
+                    if (orr.OperatorType == ExpressionType.OrElse || orr.OperatorType == ExpressionType.AndAlso)
+                    {
+                        var orElse = orr.OperatorType == ExpressionType.OrElse;
+                        var left = orr.Operands[0];
+                        var memberTargetrr = left as MemberResolveResult;
+                        bool isField = memberTargetrr != null && memberTargetrr.Member is IField &&
+                                       (memberTargetrr.TargetResult is ThisResolveResult ||
+                                        memberTargetrr.TargetResult is LocalResolveResult);
+
```
Comment (AndreyZM) [inline Compiler/Translator/Emitter/Blocks/BinaryOperatorBlock.cs]: +this.WriteOpenParentheses();

File: Compiler/Translator/Emitter/Blocks/BinaryOperatorBlock.cs [hunk 2]
```diff
@@ -52,6 +65,66 @@ protected bool ResolveOperator(BinaryOperatorExpression binaryOperatorExpression
...
+                        if (!(left is ThisResolveResult || left is TypeResolveResult || left is LocalResolveResult || left is ConstantResolveResult || isField))
+                        {
+                            this.WriteOpenParentheses();
```
Comment (AndreyZM) [inline Compiler/Translator/Emitter/Blocks/BinaryOperatorBlock.cs]: -this.WriteOpenParentheses();

**Comment by AndreyZM** [comment]
I also add coment on commits that fix this issue

**Comment by LeonidVeriga** [comment]
Hi @AndreyZM, thank you for providing one more test case!

A new issue #1892 created and fixed [here](https://github.com/bridgedotnet/Bridge/pull/1893/commits/9f243013f28a2ff3e3609a8eb197e5238cb949b7#diff-43b9dcd25c51d43f3f8d36a51ca5afacR92).

**Pull request merged**